### ISSUE TITLE
ggml_new_tensor_impl: not enough space in the context's memory pool

### ISSUE THREAD

**Issue by kmcbest** [LostRuins/koboldcpp#141]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x ] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Please provide a detailed written description of what you were trying to do, and what you expected `llama.cpp` to do.

Load a Chine Alpaca model and get it correctly generate response. This model can properly generate Chinese using [text generation webui](https://github.com/oobabooga/text-generation-webui)

And my KoboldCPP can generate English text using other models too. (I used **alpaca-7b-ggml-q4_0-lora-merged/ggml-model-q4_0.bin** from  [this link](https://wener.me/notes/ai/llama) 
# Current Behavior

Please provide a detailed written description of what `llama.cpp` did, instead.

Generation fails with a pop-up:

![image](https://user-images.githubusercontent.com/689023/236405450-e218d20e-4257-4c92-b523-47a34fbeda80.png)

# Environment and Context

Please provide detailed information about your computer setup. This is important in case the issue is not reproducible except for under certain specific conditions.

* Physical (or virtual) hardware you are using, e.g. for Linux: 
Windows 11, KoboldAPP exe 1.18. 

# Failure Information (for bugs)

Processing Prompt [BLAS] (512 / 944 tokens)ggml_new_tensor_impl: not enough space in the context's memory pool (needed 827132336, available 805306368)

# Steps to Reproduce

Please provide detailed steps for reproducing the issue. We are not sitting in front of your screen, so the more detail the better.

1. Download [the model ](https://huggingface.co/Mabbs/chinese-Alpaca-lora-7b-ggml/blob/main/ggml-model-q4_0.bin)
2. Launch KoboldApp.exe and select the model
3. Use story mode, type anything (in Chinese or in English)
4. Hit "Submit"

# Failure Logs

Please include any relevant log snippets or files. If it works under one configuration but not under another, please provide logs for both configurations and their corresponding outputs so it is easy to see where behavior changes.

Also, please try to **avoid using screenshots** if at all possible. Instead, copy/paste the console output and use [Github's markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) to cleanly format your logs for easy readability.

Example environment info:
```
Welcome to KoboldCpp - Version 1.18
For command line arguments, please refer to --help
Otherwise, please manually select ggml file:
Attempting to use OpenBLAS library for faster prompt ingestion. A compatible libopenblas will be required.
Initializing dynamic library: koboldcpp_openblas.dll
==========
Loading model: D:\#E\code\GPT\chinese-Alpaca-lora-7b-ggml-model-q4_0.bin
[Threads: 3, SmartContext: False]

---
Identified as LLAMA model: (ver 3)
Attempting to Load...
---
System Info: AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
llama.cpp: loading model from D:\#E\code\GPT\chinese-Alpaca-lora-7b-ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v1 (latest)
llama_model_load_internal: n_vocab    = 49954
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 4096
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_layer    = 32
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 11008
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 7B
llama_model_load_internal: ggml ctx size =  59.11 KB
llama_model_load_internal: mem required  = 5896.99 MB (+ 1026.00 MB per state)
llama_init_from_file: kv self size  = 1024.00 MB
Load Model OK: True
Embedded Kobold Lite loaded.
Starting Kobold HTTP Server on port 5001
Please connect to custom endpoint at http://localhost:5001
127.0.0.1 - - [05/May/2023 15:58:22] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [05/May/2023 15:58:22] "GET /api/v1/model HTTP/1.1" 200 -
127.0.0.1 - - [05/May/2023 15:58:23] "GET /api/v1/info/version HTTP/1.1" 200 -
127.0.0.1 - - [05/May/2023 15:58:23] "GET /sw.js HTTP/1.1" 404 -
127.0.0.1 - - [05/May/2023 15:58:23] "GET /manifest.json HTTP/1.1" 404 -

Input: {"n": 1, "max_context_length": 1024, "max_length": 80, "rep_pen": 1.1, "temperature": 0.72, "top_p": 0.73, "top_k": 0, "top_a": 0, "typical": 1, "tfs": 1, "rep_pen_range": 2048, "rep_pen_slope": 0.2, "sampler_order": [5, 0, 2, 3, 1, 4, 6], "prompt": ///////////EDITED//////////
Processing Prompt [BLAS] (512 / 944 tokens)ggml_new_tensor_impl: not enough space in the context's memory pool (needed 827132336, available 805306368)
----------------------------------------
Exception occurred during processing of request from ('127.0.0.1', 60680)
Traceback (most recent call last):
  File "socketserver.py", line 316, in _handle_request_noblock
  File "socketserver.py", line 347, in process_request
  File "socketserver.py", line 360, in finish_request
  File "koboldcpp.py", line 198, in __call__
  File "http\server.py", line 651, in __init__
  File "socketserver.py", line 747, in __init__
  File "http\server.py", line 425, in handle
  File "http\server.py", line 413, in handle_one_request
  File "koboldcpp.py", line 299, in do_POST
  File "koboldcpp.py", line 172, in generate
OSError: exception: access violation writing 0x0000000000000020
----------------------------------------
127.0.0.1 - - [05/May/2023 16:00:39] "POST /api/v1/generate/ HTTP/1.1" 503 -
```

**Comment by kmcbest** [comment]
Similar issues were said to be resolved:
https://github.com/ggerganov/llama.cpp/issues/153

**Comment by LostRuins** [comment]
I will merge the upstream changes soon and then you can see if that fixes it. For now, can you try with a different `--blasbatchsize` and see if that helps?

**Comment by kmcbest** [comment]
> I will merge the upstream changes soon and then you can see if that fixes it. For now, can you try with a different `--blasbatchsize` and see if that helps?

Thanks for the reply. I've tried 32, 64 and 1024, all getting the same memory error.

**Comment by LostRuins** [comment]
@kmcbest alright did some testing and it seems like the buffer sizes required for those models are larger for some reason. Can you please try the latest version and see if it works?

**Comment by kmcbest** [comment]
@LostRuins It worked, thank you!

**Comment by Ar57m** [comment]
I'm running [Cerebras 2.7](https://huggingface.co/xzuyn/Cerebras-GPT-2-Alpaca-SP-2.7B-ggml-q4_0) on Termux for Android and I'm having this same issue "not enough space in the context's memory pool", after talking a little bit it crashes and appears that message, idk why, my git clone is updated.

**Comment by LostRuins** [comment]
@Ar57m could be running out of memory. Try using `--noblas` or a smaller blas batch (e.g `--blasbatchsize 64`)

**Comment by Ar57m** [comment]
> @Ar57m could be running out of memory. Try using `--noblas` or a smaller blas batch (e.g `--blasbatchsize 64`)

I'm using this configuration:
python koboldcpp.py /storage/emulated/0/amodels/rp-chat-3b-v1-ggml-model-q4_0.bin --smartcontext --stream --noblas

For that Cerebras 2.7b model appears this after a while chatting:
Processing Prompt (8 / 8 tokens)ggml_new_tensor_impl: not enough space in the context's memory pool (needed 268695888, available 268435456)
Segmentation fault

But for [keldenl/RedPajama-INCITE-Chat-3B-v1-GGML](https://huggingface.co/keldenl/RedPajama-INCITE-Chat-3B-v1-GGML/blob/main/rp-chat-3b-v1-ggml-model-q4_0.bin) model it works fine with that configuration, maybe it's a problem with the model cerebras, but idk I'm noob.

I tested with blasbatchsize 32, but it uses too much ram, crashing a lot.

Thanks for the help @LostRuins

**Comment by LostRuins** [comment]
No worries. It's a known issue that GPT-2 based models have had issues with memory allocation, as the estimated memory per token is often incorrect. One last thing as a workaround is to use a smaller context size, which may help.
